modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/CommunicationTcpUtils.java

```diff
@@ -201,4 +205,161 @@ public static void failNode(ClusterNode nodeToFail,
+
+        /** List of predicates for each segment. */
+        private final IgnitePredicate<String>[] segmentPred = new IgnitePredicate[4];
+
+        /**
+         * Creates a new instance of IPv4 address matcher.
+         * @param interfacePattern Pattern of network interface.
+         */
+        public IPv4Matcher(String interfacePattern) {
+            super(interfacePattern);
+
+            String[] segments = networkInterface().split("\\.");
+            if (segments.length != 4)
+                throw new IllegalArgumentException("Invalid IPv4 address: " + networkInterface());
+
+            try {
+                for (int s = 0; s < segments.length; ++s) {
+                    String segment = segments[s];
+
+                    if (DIGIT_PATTERN.matcher(segment).matches()) {
+                        int seg = Integer.parseInt(segment);
+                        if (seg < 0 || seg > IPV4_MAX_OCTET_VALUE)
+                            throw new IllegalArgumentException("Invalid IPv4 address: " + networkInterface());
+
+                        segmentPred[s] = addr -> segment.equals(addr);
+                    }
+                    else if (RANGE_PATTERN.matcher(segment).matches()) {
+                        final String[] range = segment.split("-");
+                        if (range.length != 2)
+                            throw new IllegalArgumentException("Invalid IPv4 address: " + networkInterface());
+
+                        final int min = Integer.parseInt(range[0].trim());
+                        final int max = Integer.parseInt(range[1].trim());
+
+                        segmentPred[s] = addr -> {
+                            try {
+                                int seg = Integer.parseInt(addr);
+                                return seg >= min && seg <= max;
+                            } catch (NumberFormatException e) {
+                                return false;
+                            }
+                        };
+                    }
+                    else if (WILDCARD_PATTERN.equals(segments[s]))
+                        segmentPred[s] = ALWAYS_TRUE;
+                    else
+                        throw new IllegalArgumentException("Invalid IPv4 address: " + networkInterface());
+                }
+            }
+            catch (NumberFormatException e) {
+                throw new IllegalArgumentException("Invalid IPv4 address: " + networkInterface(), e);
+            }
+        }
+
+        /** {@inheritDoc} */
+        @Override boolean matches(InetAddress addr) {
+            String[] segments = addr.getHostAddress().split("\\.");
+
+            if (segments.length != 4)
+                return false;
+
+            for (int s = 0; s < segments.length; ++s) {
+                if (!segmentPred[s].apply(segments[s]))
+                    return false;
+            }
+
+            return true;
+        }
+    }
+
+    /**
+     * Represents a filter basen on blacklist of network interfaces.
```

Comment:
Probably we should mention that it returns `true` if an address is not blacklisted, this is not obvious

Reply:
This is already mentioned here:
```
        /**
         * Returns {@code true} if the given address is not in the blacklist and {@code false} otherwise.
         *
         * @param inetAddress Address to be checked
         * @return {@code true} if the given address is not in the blacklist and {@code false} otherwise.
         */
        @Override public boolean apply(InetAddress inetAddress)
```
Anyway, it is worth mentioning in the javadoc of the class level too.